feat: add FiNER Wallet Support

prih opened the pull request:
# Description

Add FiNER Wallet App support.

FiNER Wallet implements an API almost identical to near-wallet over Deep linking on IOS and Android.

# Checklist:
<!-- CHECKLIST_TYPE: ALL -->
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
<!-- /CHECKLIST_TYPE -->

# Type of change. This type of change is the main reason for the PR.
<!-- CHECKLIST_TYPE: ONE -->
- [x] FEATURE - a PR of this type introduces a new feature.
<!-- /CHECKLIST_TYPE -->

# Breaking changes
<!-- CHECKLIST_TYPE: ONE -->
<!-- SPECIFY BREAKING CHANGES AS A ONE-LINER -->
- [x] NO BREAKING CHANGE - this PR doesn't contain any breaking changes and it's backwards compatible
<!-- /CHECKLIST_TYPE -->

kujtimprenkuSQA reviewed (commented):
FiNer Wallet needs to be removed from the `default-wallets` package.

When trying to sign in I see this error, I am using an iOS device iPhone X :
![MicrosoftTeams-image (5)](https://user-images.githubusercontent.com/95851345/205059211-49e5827b-3620-479a-a8bf-ee21cd34d641.png)

prih commented:
> FiNer Wallet needs to be removed from the `default-wallets` package.
> 
> When trying to sign in I see this error, I am using an iOS device iPhone X :
> 
> ![MicrosoftTeams-image (5)](https://user-images.githubusercontent.com/95851345/205060060-38ddc152-0fba-4284-9e25-2b2b3853858d.png)
> 
> I have imported `kujtimprenku.testnet` via mnemonic.

you need to separately enable the near-testnet blockchain, for this you need to go to the Settings - Developer mode (turn on) section, after that Settings - Wallet manage - Manage assets - Testnet Near (turn on)

this is separate for security reasons, there are cases of phishing and the user is sent fake assets on the testnet network under the guise of mainnet

prih commented on examples/react/project.json:
```diff
@@ -77,6 +77,11 @@
             "input": "packages/opto-wallet/assets/",
             "output": "assets/"
+          },
+          {
+            "glob": "**/*",
+            "input": "packages/finer-wallet/assets/",
+            "output": "assets/"
           }
```
done

prih commented on packages/default-wallets/src/lib/default-wallets.ts:
```diff
@@ -1,9 +1,10 @@
   Array<WalletModuleFactory>
 > => {
-  return [setupMyNearWallet(), setupLedger()];
+  return [setupMyNearWallet(), setupLedger(), setupFinerWallet()];
```
ok, done

who decides what wallets should be there? why MyNearWallet ?

kujtimprenkuSQA replied:
It's mentioned here in the contributing docs: https://github.com/near/wallet-selector/blob/main/CONTRIBUTING.md#default-wallet-list-criteria

prih replied:
ok thx

kujtimprenkuSQA commented:
Hey, @prih thank you for addressing the comments.

We have been trying to test FiNer Wallet on different iOS devices but we had no luck seeing the full functionality working properly.

In iPhone X when enabling Developer mode and then Near Testnet it just shows "Syncing..." all the time it never changes from this state.

In some other devices even after successfully enabling Near Testnet and importing an account when we try to signIn and accept it on the wallet after we get redirected back to the Guest Book example it does not show as signedIn.

I have forked your work here: https://github.com/kujtimprenkuSQA/wallet-selector/tree/dev
Here's a published version of it in gh-pages to make testing easier: https://kujtimprenkusqa.github.io/wallet-selector/

Here's a video that shows the issue we're having:

https://user-images.githubusercontent.com/95851345/205150582-587cfa82-63a6-4bca-ad9d-704fe4665465.MOV

prih commented:
> Hey, @prih thank you for addressing the comments.
> 
> We have been trying to test FiNer Wallet on different iOS devices but we had no luck seeing the full functionality working properly.
> 
> In iPhone X when enabling Developer mode and then Near Testnet it just shows "Syncing..." all the time it never changes from this state.
> 
> In some other devices even after successfully enabling Near Testnet and importing an account when we try to signIn and accept it on the wallet after we get redirected back to the Guest Book example it does not show as signedIn.
> 
> I have forked your work here: https://github.com/kujtimprenkuSQA/wallet-selector/tree/dev Here's a published version of it in gh-pages to make testing easier: https://kujtimprenkusqa.github.io/wallet-selector/
> 
> Here's a video that shows the issue we're having:
> 
>  IMG_6320.MOV


https://user-images.githubusercontent.com/3285877/205161351-9ff75e73-ab5b-48f3-b102-387f9d9857cf.MP4

prih commented:
> Hey, @prih thank you for addressing the comments.
> 
> We have been trying to test FiNer Wallet on different iOS devices but we had no luck seeing the full functionality working properly.
> 
> In iPhone X when enabling Developer mode and then Near Testnet it just shows "Syncing..." all the time it never changes from this state.
> 
> In some other devices even after successfully enabling Near Testnet and importing an account when we try to signIn and accept it on the wallet after we get redirected back to the Guest Book example it does not show as signedIn.
> 
> I have forked your work here: https://github.com/kujtimprenkuSQA/wallet-selector/tree/dev Here's a published version of it in gh-pages to make testing easier: https://kujtimprenkusqa.github.io/wallet-selector/
> 
> Here's a video that shows the issue we're having:
> 
>  IMG_6320.MOV

I don’t know what the problem could be, there is one thought, perhaps the transaction has not yet reached the node with which your page works, and therefore, immediately after the redirect, it does not see the new state....
this is a huge problem for such a login method, but there is nothing we can do about it...

about the long synchronization ... what version of Finer is installed? where are you geographically from? US or EU?

prih commented:
> Hey, @prih thank you for addressing the comments.
> 
> We have been trying to test FiNer Wallet on different iOS devices but we had no luck seeing the full functionality working properly.
> 
> In iPhone X when enabling Developer mode and then Near Testnet it just shows "Syncing..." all the time it never changes from this state.
> 
> In some other devices even after successfully enabling Near Testnet and importing an account when we try to signIn and accept it on the wallet after we get redirected back to the Guest Book example it does not show as signedIn.
> 
> I have forked your work here: https://github.com/kujtimprenkuSQA/wallet-selector/tree/dev Here's a published version of it in gh-pages to make testing easier: https://kujtimprenkusqa.github.io/wallet-selector/
> 
> Here's a video that shows the issue we're having:
> 
>  IMG_6320.MOV

it is also possible that the problem is in the rpc that the page is using if it is rpc.testnet.near.org - it never works correctly... we have used it before, often the nodes are behind in time, often 50x errors, often the indexer lags and loses transactions, when we raised our nodes and indexer, everything began to work as it should.

kujtimprenkuSQA commented:
>about the long synchronization ... what version of Finer is installed? where are you geographically from? US or EU?

I have installed version `0.3.7` of FiNer Wallet and I am located in EU.

>it is also possible that the problem is in the rpc that the page is using if it is rpc.testnet.near.org - it never works correctly... we have used it before, often the nodes are behind in time, often 50x errors, often the indexer lags and loses transactions, when we raised our nodes and indexer, everything began to work as it should.

We use https://rpc.testnet.near.org/ in the guest book example and it works fine for other wallets to `signIn`, `signAndSendTransaction(s)`.

prih commented:
> > about the long synchronization ... what version of Finer is installed? where are you geographically from? US or EU?
> 
> I have installed version `0.3.7` of FiNer Wallet and I am located in EU.
> 
> > it is also possible that the problem is in the rpc that the page is using if it is rpc.testnet.near.org - it never works correctly... we have used it before, often the nodes are behind in time, often 50x errors, often the indexer lags and loses transactions, when we raised our nodes and indexer, everything began to work as it should.
> 
> We use https://rpc.testnet.near.org/ in the guest book example and it works fine for other wallets to `signIn`, `signAndSendTransaction(s)`.

I checked several times, everything works correctly, I'll try on different iPhone models... but I don't think this is the case

kujtimprenkuSQA commented:
A bit more feedback:

The issue with syncing was happening only on this account id for some reason `kujtimprenku.testnet`.

Tested importing `prenku.testnet` and `kujtim.testnet` I could import them successfully and the testnet network was enabled and synced.

While testing the above guest book example I noticed a few things:

- When I visit the [example](https://kujtimprenkusqa.github.io/wallet-selector/) and try to sign in with FiNer Wallet from Chrome browser after I confirm the sign in in the wallet I get redirected in Safari and not in Chrome (of course it shows as not signed in).

- When I try the sign in from Safari after confirming the signIn in the wallet when I get redirected a new tab is being created and I don't get redirected to the tab I initially clicked the FiNer Wallet option.
- `signAndSendTransaction`(s) with donation/deposit is not working.

AmmarHumackicSQA commented:
> While testing the above guest book example I noticed a few things:
> 
> * When I visit the [example](https://kujtimprenkusqa.github.io/wallet-selector/) and try to sign in with FiNer Wallet from Chrome browser after I confirm the signIn in the wallet I get redirected in Safari and not in Chrome (of course it shows as not signed in).
> * When I try the sign in from Safari after confirming the signIn in the wallet when I get redirected a new tab is being created and I don't get redirected to the tab I initially clicked the FiNer Wallet option.
> * `signAndSendTransaction`(s) with donation/deposit is not working shows a [404 not found](https://finerwallet.io/sign?transactions=DgAAAGt1anRpbS50ZXN0bmV0AKjIHrxCEJZ9rGoz%2B73ekHH0x39pcsG5cRyl75CecqtrB8TCNe5JAAASAAAAZ3Vlc3QtYm9vay50ZXN0bmV0z3d3n1uUAMyyIUloQsNpHdC97M0Bh8POUicF30p40wQBAAAAAgoAAABhZGRNZXNzYWdlEAAAAHsidGV4dCI6IlRlc3QxIn0A4FfrSBsAAAAAQLK6yeAZHgIAAAAAAAA%3D&callbackUrl=https%3A%2F%2Fkujtimprenkusqa.github.io%2Fwallet-selector%2F).


Yeah, same for me 👆

prih commented:
> A bit more feedback:
> 
> The issue with syncing was happening only on this account id for some reason `kujtimprenku.testnet` (so the issue is not with the device).
> 
> Tested importing `prenku.testnet` and `kujtim.testnet` I could import them successfully and the testnet network was enabled and synced.
> 
> While testing the above guest book example I noticed a few things:
> 
> * When I visit the [example](https://kujtimprenkusqa.github.io/wallet-selector/) and try to sign in with FiNer Wallet from Chrome browser after I confirm the signIn in the wallet I get redirected in Safari and not in Chrome (of course it shows as not signed in).
> * When I try the sign in from Safari after confirming the signIn in the wallet when I get redirected a new tab is being created and I don't get redirected to the tab I initially clicked the FiNer Wallet option.
> * `signAndSendTransaction`(s) with donation/deposit is not working shows a [404 not found](https://finerwallet.io/sign?transactions=DgAAAGt1anRpbS50ZXN0bmV0AKjIHrxCEJZ9rGoz%2B73ekHH0x39pcsG5cRyl75CecqtrB8TCNe5JAAASAAAAZ3Vlc3QtYm9vay50ZXN0bmV0z3d3n1uUAMyyIUloQsNpHdC97M0Bh8POUicF30p40wQBAAAAAgoAAABhZGRNZXNzYWdlEAAAAHsidGV4dCI6IlRlc3QxIn0A4FfrSBsAAAAAQLK6yeAZHgIAAAAAAAA%3D&callbackUrl=https%3A%2F%2Fkujtimprenkusqa.github.io%2Fwallet-selector%2F).

yes, confirm, we are investigating...

kujtimprenkuSQA commented:
You might have come to the same conclusion, I think the reason `signAndSendTransaction(s)` is returning a 404 not found for FiNer Wallet is because of this [function](https://github.com/near/near-api-js/blob/master/packages/near-api-js/src/wallet-account.ts#L205-L217) which `near-api-js` uses when you add a deposit to a transaction.

Especially this line of code:
```ts
// this._walletBaseUrl = "https://finerwallet.io/near-wallet-testnet";
const newUrl = new URL('sign', this._walletBaseUrl);
```

Because when it gets executed this is going to be the result:

![newUrl](https://user-images.githubusercontent.com/95851345/205345837-ed7ee0b5-d1a1-4ee4-9cc9-52fb66ac8554.PNG)

In your wallet maybe you expect this to be something like:
https://finerwallet.io/near-wallet-testnet/sign?transactions=transactionHere

But the function above constructs the URL like this: 
https://finerwallet.io/sign?transactions=transactionHere

The fix for this could be to deploy your wallet on a subdomain: https://near-wallet-testnet.finerwallet.io (this is the approach all other browser wallets follow)

![newUrl2](https://user-images.githubusercontent.com/95851345/205347818-ffe01fd3-fa5a-4e32-b58b-39ee935a23a5.PNG)

prih commented:
> You might have come to the same conclusion, I think the reason `signAndSendTransaction(s)` is returning a 404 not found for FiNer Wallet is because of this [function](https://github.com/near/near-api-js/blob/master/packages/near-api-js/src/wallet-account.ts#L205-L217) which `near-api-js` uses when you add a deposit to a transaction.
> 
> Especially this line of code:
> 
> ```ts
> // this._walletBaseUrl = "https://finerwallet.io/near-wallet-testnet";
> const newUrl = new URL('sign', this._walletBaseUrl);
> ```
> 
> Because when it gets executed this is going to be the result:
> 
> ![newUrl](https://user-images.githubusercontent.com/95851345/205345837-ed7ee0b5-d1a1-4ee4-9cc9-52fb66ac8554.PNG)
> 
> In your wallet maybe you expect this to be something like: https://finerwallet.io/near-wallet-testnet/sign?transactions=transactionHere
> 
> But the function above constructs the URL like this: https://finerwallet.io/sign?transactions=transactionHere
> 
> The fix for this could be to deploy your wallet on a subdomain: https://near-wallet-testnet.finerwallet.io (this is the approach all other browser wallets follow)
> 
> ![newUrl2](https://user-images.githubusercontent.com/95851345/205347818-ffe01fd3-fa5a-4e32-b58b-39ee935a23a5.PNG)

yes, thanks, we are solving this problem right now, and will soon update the mobile application and the integration code with wallet-selector

about deployment) we have a mobile application, all communication pages are just intermediate pages

prih commented:
> A bit more feedback:
> 
> The issue with syncing was happening only on this account id for some reason `kujtimprenku.testnet` (so the issue is not with the device).
> 
> Tested importing `prenku.testnet` and `kujtim.testnet` I could import them successfully and the testnet network was enabled and synced.
> 
> While testing the above guest book example I noticed a few things:
> 
> * When I visit the [example](https://kujtimprenkusqa.github.io/wallet-selector/) and try to sign in with FiNer Wallet from Chrome browser after I confirm the signIn in the wallet I get redirected in Safari and not in Chrome (of course it shows as not signed in).
> * When I try the sign in from Safari after confirming the signIn in the wallet when I get redirected a new tab is being created and I don't get redirected to the tab I initially clicked the FiNer Wallet option.
> * `signAndSendTransaction`(s) with donation/deposit is not working shows a [404 not found](https://finerwallet.io/sign?transactions=DgAAAGt1anRpbS50ZXN0bmV0AKjIHrxCEJZ9rGoz%2B73ekHH0x39pcsG5cRyl75CecqtrB8TCNe5JAAASAAAAZ3Vlc3QtYm9vay50ZXN0bmV0z3d3n1uUAMyyIUloQsNpHdC97M0Bh8POUicF30p40wQBAAAAAgoAAABhZGRNZXNzYWdlEAAAAHsidGV4dCI6IlRlc3QxIn0A4FfrSBsAAAAAQLK6yeAZHgIAAAAAAAA%3D&callbackUrl=https%3A%2F%2Fkujtimprenkusqa.github.io%2Fwallet-selector%2F).

we have a mobile application, it is not a web wallet, it does not work in a browser, and therefore we cannot return to the desired page, or the desired browser ... but we are now deciding how best to proceed
